# **Task 2 - Warm Up Vaccine Data**

In the tasks below, we use the vaccine dataset and the sheet Vaccinated municipality and age.
W will use Plotly express to draw the diagrams below for the tasks that require diagrams and use Pandas
to answer questions about the data set.

In [22]:
import pandas as pd

In [23]:
data = pd.read_excel("Data/Folkhalsomyndigheten_Covid19_Vaccine.xlsx", sheet_name="Vaccinerade kommun och ålder")
df = pd.DataFrame(data)
df.head(3)

Län        Län_namn  Kommun     Kommun_namn  Ålder  Befolkning  \
0    1  Stockholms län     114  Upplands Väsby  12-15        2422   
1    1  Stockholms län     114  Upplands Väsby  16-17        1203   
2    1  Stockholms län     114  Upplands Väsby  18-29        6692   

   Antal minst 1 dos  Antal minst 2 doser  Antal 3 doser  Antal 4 doser  \
0               1206                 1046            NaN            NaN   
1                839                  755            NaN            NaN   
2               4887                 4469         1959.0            NaN   

   Andel minst 1 dos  Andel minst 2 doser  Andel 3 doser  Andel 4 doser  
0           0.497936             0.431874            NaN            NaN  
1           0.697423             0.627598            NaN            NaN  
2           0.730275             0.667812       0.292738            NaN

### **a)** How many counties are represented in the data set?

We use the .value_counts() method to count aggregate a list of all unique values () we have in the column representing county name, then we count the number of values in this list using len().

In [24]:
print(f"The number of counties in the dataset is: {len(df['Län_namn'].value_counts())}.\nIt is the following counties (followed by number of rows):\n{df['Län_namn'].value_counts()}")

The number of counties in the dataset is: 21.
It is the following counties (followed by number of rows):
Län_namn
Västra Götalands län    490
Skåne län               330
Stockholms län          260
Värmlands län           160
Västerbottens län       150
Dalarnas län            150
Norrbottens län         140
Jönköpings län          130
Östergötlands län       130
Kalmar län              120
Örebro län              120
Västmanlands län        100
Gävleborgs län          100
Södermanlands län        90
Uppsala län              80
Kronobergs län           80
Jämtlands län            80
Västernorrlands län      70
Hallands län             60
Blekinge län             50
Gotlands län             10
Name: count, dtype: int64


In the above code we can see that all 21 of Swedens counties are represented in the data set.

Source: [SKR](https://skr.se/skr/tjanster/kommunerochregioner/faktakommunerochregioner.1022.html)

### **b)** How many municipalities are represented in the data set?

Here we use the same approach as seen above, but this time counting municipalities.

In [25]:
print(f"The number of municipalities in the dataset is: {len(df['Kommun_namn'].value_counts())}.")

The number of municipalities in the dataset is: 290.


In the above code we can see that all 290 of Swedens municipalities are represented in the dataset.

Source: [SKR](https://skr.se/skr/tjanster/kommunerochregioner/faktakommunerochregioner.1022.html)

### **c)** How large is the population represented in the data set?


Down below we calculating the sum of our column with values for population using .sum() method.

In [26]:
print(f"The total sum of the population represented in the data set is: {df['Befolkning'].sum()}.")

The total sum of the population represented in the data set is: 9092790.


According to [SCB](https://www.scb.se/hitta-statistik/statistik-efter-amne/befolkning/befolkningens-sammansattning/befolkningsstatistik/pong/tabell-och-diagram/folkmangd-och-befolkningsforandringar---manad-kvartal-och-halvar/befolkningsstatistik-2023/) (Statistiska Centralbyrån) the total population of Sweden in oktober 2022 where 
10,519,592 but in our data set only 9,092,790 is represented. In the code below we can see that there is no NaN/Null values in the 'Befolkning column' so the reason in not because of values missing.

In [27]:
df.isnull().sum()

Län                       0
Län_namn                  0
Kommun                    0
Kommun_namn               0
Ålder                     0
Befolkning                0
Antal minst 1 dos         0
Antal minst 2 doser       0
Antal 3 doser           580
Antal 4 doser          2030
Andel minst 1 dos         0
Andel minst 2 doser       0
Andel 3 doser           580
Andel 4 doser          2030
dtype: int64

### **d)** Calculate how many children under the age of 18 there are in Sweden based on this data set. 

First I need to find the unique values in the data set using the .unique() method on the column 'Ålder' (age).

In [28]:
df['Ålder'].unique()

array(['12-15', '16-17', '18-29', '30-39', '40-49', '50-59', '60-69',
       '70-79', '80-89', '90 eller äldre'], dtype=object)

The above code indicates that the age groups for children includes children aged 12 to 17 years. Down below we include the two groups below 18 in our .query() method and makning a new dataframe called 'children_12_to_17'. Then we aggregate the total sum of the population column from our new dataframe.

In [ ]:
df_stockholm_västra_götaland = df.query("Län_namn == 'Stockholms län' | Län_namn == 'Västra Götalands län'").copy()

new_df_sthlm_vgr = pd.DataFrame()
new_df_sthlm_vgr['Share 1 dose'] = (df_stockholm_västra_götaland.groupby('Län_namn')['Antal minst 1 dos'].sum() / df_stockholm_västra_götaland.groupby('Län_namn')['Befolkning'].sum())*100
new_df_sthlm_vgr['Share 2 doses'] = (df_stockholm_västra_götaland.groupby('Län_namn')['Antal minst 2 doser'].sum() / df_stockholm_västra_götaland.groupby('Län_namn')['Befolkning'].sum())*100
new_df_sthlm_vgr['Share 3 doses'] = (df_stockholm_västra_götaland.groupby('Län_namn')['Antal 3 doser'].sum() / df_stockholm_västra_götaland.groupby('Län_namn')['Befolkning'].sum())*100
new_df_sthlm_vgr['Share 4 doses'] = (df_stockholm_västra_götaland.groupby('Län_namn')['Antal 4 doser'].sum() / df_stockholm_västra_götaland.groupby('Län_namn')['Befolkning'].sum())*100

fig = px.bar(new_df_sthlm_vgr, 
             x=new_df_sthlm_vgr.index, 
             y=['Share 1 dose', 'Share 2 doses', 'Share 3 doses', 'Share 4 doses'],
             labels=dict(Län_namn= 'County', value='Percent', variable='Number of doses'),
             title='Vaccination Percent for 1, 2, 3 and 4 Doses in Stockholm and Västra Götalands Counties',
             barmode='group',
             template='simple_white')

fig.show()
fig.write_html("Visualizations/dose_1_to_4_STHLM_VGR.html")

In [29]:
children_12_to_17 = df.query("Ålder == ['12-15', '16-17']")['Befolkning'].sum()
print(f"{children_12_to_17 = }")

children_12_to_17 = 745370


Let's consider children under 12 in the year 2022. If we assume that a similar number of children were born after 2010 as the number of individuals aged 12-17 in our dataset, it means one-third of the total children are represented in our data.

To find the remaining children (those aged 0-11 in 2022), we can double the number of children represented in our dataset. Adding these missing children (aged 0-11) to the total population in our dataset (9,092,790) brings us to a value closer to the SCB data mentioned in question c) (10,519,592). Specifically, our calculation results in a total population of 10,583,530, aligning more closely with the SCB data.

In [30]:
children_0_to_11 = children_12_to_17 * 2
total_population = children_0_to_11 + df['Befolkning'].sum()
total_population

10583530

In [31]:
minors_sweden = children_0_to_11 + children_12_to_17
minors_sweden

2236110

As seen above the total number of children younger than 18 years would be 2,236,110 and this divided by the total population 10583530 (total population in dataset calculation) would give us the result that approximately every fifth person in Sweden were below 18 years old in the autum of 2022. The answer is presented in percent for better readability.

In [32]:
print(f"The children below 18 were {(minors_sweden/total_population) * 100:.2f}% in the autum of 2022 according to our data set.")

The children below 18 were 21.13% in the autum of 2022 according to our data set.


### **e)** Draw a diagram showing the age distribution of Sweden's population.

In [33]:
import plotly.express as px

To plot the actual population in Sweden in the year of 2022 we looked for data on SCB (Statistics Sweden). We turn this into a dataframe using Pandas, and then rename the columns to 'Ålder' (age) and 'Antal' (Amount) saved in the variable pop_sweden_2022.

In [34]:
pop_sweden_2022 = pd.read_excel("Data/swedish_pop.xlsx") # Collected data here: https://www.statistikdatabasen.scb.se/sq/141883
pop_sweden_2022.columns=['Ålder', 'Antal']
pop_sweden_2022

Ålder    Antal
0     0-4 år   576367
1    5-14 år  1252726
2   15-24 år  1188337
3   25-34 år  1447457
4   35-44 år  1342361
5   45-54 år  1313360
6   55-64 år  1253811
7   65-74 år  1067123
8   75-84 år   805998
9   85-94 år   250340
10    95+ år    23676

Down below we can see the population in Sweden in 2022. Most people in Sweden where by then in their late twenties or early thirties.

We plot the data using plotly express histogram and change the label names to age and amount, set a title, redesign it using template, adding colors for the bins and adding hover texts. In the end we save our plot in a html file using the write_html() method.

In [35]:
fig = px.histogram(
    pop_sweden_2022, 
    x='Ålder', 
    y='Antal', 
    title='Age Distribution of Sweden\'s Population 2022', 
    template='simple_white',
    color='Ålder',
    labels=dict(Ålder='Age', Antal='Amount',),
    hover_name='Ålder'
)

fig.update_yaxes(title_text='Amount')
fig.show()
fig.write_html("Visualizations/age_distributions_swedish_pop.html")

### **f)** Draw bar graphs for proportion with at least 1 dose per county, proportion of two doses per county and proportion of three doses per county.

The code below groups the dataset by county ('Län_namn'). It calculates the vaccination percentages for three doses ('Share 1 dose', 'Share 2 doses', and 'Share 3 doses') by summing the doses administered ('Antal minst 1 dos', 'Antal minst 2 doser', 'Antal 3 doser') and dividing by the total population ('Befolkning') of each county. The calculated percentages are multiplied by 100 to convert them into percentages.

We Plotly Express to create a grouped bar chart. The chart visualizes the vaccination rates for 1, 2, and 3 doses across different counties. Counties are shown on the x-axis, and the y-axis represents the vaccination percentages. The chart is displayed and saved as an HTML file for further reference.

In summary, the code analyzes vaccination data for various counties, calculates the vaccination rates for different doses, and presents the results in an interactive grouped bar chart for easy visualization and analysis.

In [68]:
new_df_all_counties = pd.DataFrame()
new_df_all_counties['Share 1 dose'] = (df.groupby('Län_namn')['Antal minst 1 dos'].sum() / df.groupby('Län_namn')['Befolkning'].sum())*100
new_df_all_counties['Share 2 doses'] = (df.groupby('Län_namn')['Antal minst 2 doser'].sum() / df.groupby('Län_namn')['Befolkning'].sum())*100
new_df_all_counties['Share 3 doses'] = (df.groupby('Län_namn')['Antal 3 doser'].sum() / df.groupby('Län_namn')['Befolkning'].sum())*100

fig = px.bar(new_df_all_counties, x=new_df_all_counties.index, 
             y=['Share 1 dose', 'Share 2 doses', 'Share 3 doses'],
             labels={'value': 'Percent', 'variable': 'Number of doses', 'Län_namn': 'County'},
             title='Share Vaccination percent for 1, 2, and 3 Doses by County',
             barmode='group',
             template='simple_white')

fig.update_xaxes(tickangle=45, categoryorder='total ascending')
fig.add_hline(y=83, line_width=2, line_dash='dot')
fig.write_html("Visualizations/dose_1_2_3_total_pop.html")
fig.show()

Important notes to make here is that not all counties have provided data for dose three, as seen below. But we can draw the conclusion that all counties where are quite close in shares of dose 1. In bottom we find Skåne and Stockholm, for dose 1, 2 and 3. The largest share of the population that received dose 1 and 2 where in Västerbottens county and 3 doses on Gotland.

In [37]:
df.isnull().sum()

Län                       0
Län_namn                  0
Kommun                    0
Kommun_namn               0
Ålder                     0
Befolkning                0
Antal minst 1 dos         0
Antal minst 2 doser       0
Antal 3 doser           580
Antal 4 doser          2030
Andel minst 1 dos         0
Andel minst 2 doser       0
Andel 3 doser           580
Andel 4 doser          2030
dtype: int64

### **g)** Take Västra Götaland County and Stockholm County and draw a diagram of the proportion of those vaccinated with at least 1 dose, two doses, three doses and four doses.

Down below we've taken a similar approch as above but making a new dataframe containing only data for Stockholm and Västra Götaland. We've also added the column 'Andel 4 doser' (Share 4 doses).

In [52]:
df_stockholm_västra_götaland = df.query("Län_namn == 'Stockholms län' | Län_namn == 'Västra Götalands län'").copy()

new_df_sthlm_vgr = pd.DataFrame()
new_df_sthlm_vgr['Share 1 dose'] = (df_stockholm_västra_götaland.groupby('Län_namn')['Antal minst 1 dos'].sum() / df_stockholm_västra_götaland.groupby('Län_namn')['Befolkning'].sum())*100
new_df_sthlm_vgr['Share 2 doses'] = (df_stockholm_västra_götaland.groupby('Län_namn')['Antal minst 2 doser'].sum() / df_stockholm_västra_götaland.groupby('Län_namn')['Befolkning'].sum())*100
new_df_sthlm_vgr['Share 3 doses'] = (df_stockholm_västra_götaland.groupby('Län_namn')['Antal 3 doser'].sum() / df_stockholm_västra_götaland.groupby('Län_namn')['Befolkning'].sum())*100
new_df_sthlm_vgr['Share 4 doses'] = (df_stockholm_västra_götaland.groupby('Län_namn')['Antal 4 doser'].sum() / df_stockholm_västra_götaland.groupby('Län_namn')['Befolkning'].sum())*100

fig = px.bar(new_df_sthlm_vgr, 
             x=new_df_sthlm_vgr.index, 
             y=['Share 1 dose', 'Share 2 doses', 'Share 3 doses', 'Share 4 doses'],
             labels=dict(Län_namn= 'County', value='Percent', variable='Number of doses'),
             title='Vaccination Percent for 1, 2, 3 and 4 Doses in Stockholm and Västra Götalands Counties',
             barmode='group',
             template='simple_white')

fig.show()
fig.write_html("Visualizations/dose_1_to_4_STHLM_VGR.html")

From the above we can draw the conclusion that Västra Götaland has a higher percentage of vaccinated regarding all four categories.